src/server/bitops_family.cc

```diff
@@ -183,60 +287,203 @@ bool SetBitValue(uint32_t offset, bool bit_value, std::string* entry) {
+  bool old_value = false;
 
-  std::pair<PrimeIterator, bool> add_res;
-  try {
-    add_res = db_slice.AddOrFind(index_, key);
-  } catch (const std::bad_alloc&) {
-    return OpStatus::OUT_OF_MEMORY;
+  ElementAccess element_access{key, index_};
+  auto find_res = element_access.Find(shard_);
+
+  if (find_res != OpStatus::OK) {
+    return find_res;
   }
-  bool old_value = false;
-  PrimeIterator& it = add_res.first;
-  bool added = add_res.second;
-  auto UpdateBitMapValue = [&](std::string_view value) {
-    db_slice.PreUpdate(index, it);
-    it->second.SetString(value);
-    db_slice.PostUpdate(index, it, key, !added);
-  };
 
-  if (added) {  // this is a new entry in the "table"
+  if (element_access.IsNewEntry()) {
     std::string new_entry(GetByteIndex(offset) + 1, 0);
     old_value = SetBitValue(offset, bit_value, &new_entry);
-    UpdateBitMapValue(new_entry);
+    element_access.Commit(new_entry);
   } else {
-    if (it->second.ObjType() != OBJ_STRING) {
-      return OpStatus::WRONG_TYPE;
-    }
     bool reset = false;
-    std::string existing_entry{GetString(shard_, it->second)};
+    std::string existing_entry{element_access.Value()};
     if ((existing_entry.size() * OFFSET_FACTOR) <= offset) {  // need to resize first
       existing_entry.resize(GetByteIndex(offset) + 1, 0);
       reset = true;
     }
     old_value = SetBitValue(offset, bit_value, &existing_entry);
     if (reset || old_value != bit_value) {  // we made a "real" change to the entry, save it
-      UpdateBitMapValue(existing_entry);
+      element_access.Commit(existing_entry);
     }
   }
   return old_value;
 }
 
+// ---------------------------------------------------------
+
+std::string RunBitOperationOnValues(std::string_view op, BitOpValues values) {
+  // This function accept an operation (either OR, XOR, NOT or OR), and run bit operation
+  // on all the values we got from the database. Note that in case that one of the values
+  // is shorter than the other it would return a 0 and the operation would continue
+  // until we ran the longest value. The function will return the resulting new value
+  std::size_t max_len = 0;
+  std::size_t max_len_index = 0;
+
+  const auto BitOperation = [&]() {
+    if (op == OR_OP_NAME) {
+      BitOpString operation{max_len, OrOp, SkipOr};
+      std::string default_str{values[max_len_index].get_ref()};
+      return operation.Run(std::move(values), std::string(max_len, 0));
+    } else if (op == XOR_OP_NAME) {
+      BitOpString operation{max_len, XorOp, SkipXor};
+      return operation.Run(std::move(values), std::string(max_len, 0));
+    } else if (op == AND_OP_NAME) {
+      BitOpString operation{max_len, AndOp, SkipAnd};
+      return operation.Run(std::move(values), std::string(max_len, 0));
+    } else if (op == NOT_OP_NAME) {
+      return BitOpNotString(values[0].take());
+    } else {
+      return std::string{};
+    }
+  };
+  if (values.empty()) {
+    return std::string{};
```

Comment:
it's not true imho.
command spec prohibits running the operation without source keys.
if values can not be empty, pls add DCHECK instead of returning.